doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb

```diff
@@ -0,0 +1,917 @@
+    "    resnet18 = initialize_model()\n",
+    "    return checkpoint.get_model(model=resnet18)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define the Training Loop\n",
+    "\n",
+    "The `train_loop_per_worker` function defines the finetuning procedure for each worker.\n",
+    "\n",
+    "**1. Load dataset shard for each worker**:\n",
+    "- A ray trainer will take a dictionary of ray datasets as input. One can accessed these data by `session.get_dataset_shard(DATASET_KEY)` in the workers.\n",
+    "- Only the dataset with key \"train\" will be split into multiple shards, while all the others will remain the same. \n",
+    "- One can use {meth}`iter_torch_batches <ray.data.Dataset.iter_torch_batches>` to iterate the datasets with automatic tensor batching. If you need more flexible customized batching function, please refer to our lower-level api {meth}`iter_batches <ray.data.Dataset.iter_batches>`.\n",
+    "\n",
+    "**2. Prepare your model**:\n",
+    "- `train.torch.prepare_model` will prepares the model for distributed training. Under the hood, it converts your torch model to `DistributedDataParallel` model, and synchronizes the gradients and buffers across all workers.\n",
+    "\n",
+    "**3. Report metrics and checkpoint**:\n",
+    "- `session.report` will gather the metrics from each worker and save them into log files.\n",
+    "- You don't have to save checkpoints manually with `torch.save()`, `session.report()` will help you sync checkpoints to local/cloud storage.\n",
+    "- The best checkpoints will be saved according to the specified `checkpoint_score_attribute` in {class}`CheckpointConfig <ray.air.config.CheckpointConfig>`. Here we only save the best model with highest validation accuracy."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ray.train as train\n",
+    "from ray.air import session\n",
+    "from ray.train.torch import TorchCheckpoint\n",
+    "\n",
+    "\n",
+    "def evaluate(logits, labels):\n",
+    "    _, preds = torch.max(logits, 1)\n",
+    "    corrects = torch.sum(preds == labels).item()\n",
+    "    return corrects\n",
+    "\n",
+    "\n",
+    "def train_loop_per_worker(configs):\n",
+    "    # Prepare dataloader for each worker\n",
+    "    datasets = dict()\n",
+    "    datasets[\"train\"] = session.get_dataset_shard(\"train\")\n",
+    "    datasets[\"val\"] = session.get_dataset_shard(\"val\")\n",
+    "\n",
+    "    # Calculate the batch size for a single worker\n",
+    "    worker_batch_size = configs[\"batch_size\"] // session.get_world_size()\n",
+    "\n",
+    "    device = \"cpu\"\n",
+    "    if torch.cuda.is_available():\n",
+    "        device = \"cuda:0\"\n",
+    "\n",
+    "    # Prepare DDP Model, optimizer, and loss function\n",
+    "    model = initialize_model_from_ckpt()\n",
```

Comment:
Might want to also load epoch/optimizer state if they are in the checkpoint dict.

Reply:
Are we making this demo overly complicated? I think this demo just aim to show how to finetune a pretrained model, normally from scratch. 
Resuming training from checkpoint looks like another topic, fault tolerance?

Reply:
Yeah, let's not load epoch/optimizer state. But I think we should go through the `session.get_checkpoint` path, since that's the `resume_from_checkpoint` API.